### PR TITLE
coala_main.py: Provides options with -A tag

### PR DESCRIPTION
**Pull request by Naveenaidu**
This allows selection of the actions when the -A
tag is used.The actions can now be selected with
the appropriate number or character

Fixes https://github.com/coala/coala/pull/5166

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/coala_main.py [hunk 1]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
     """
     all_actions_possible = provide_all_actions()
     apply_single = None
+
```
Comment (ishanSrt) [inline coalib/coala_main.py]: unnecessary change

File: coalib/coala_main.py [hunk 2]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
...
+                    index = str('(') + str(index + 1) + str(')')
+                    if choice in action or choice in index:
                         apply_single = action
                         break
                 if apply_single:
                     break
                 console_printer.print(format_lines(
                                     'Please enter a valid letter.',
                                     symbol='['))
-
```
Comment (ishanSrt) [inline coalib/coala_main.py]: Please don't add/delete lines unnecessarily

File: coalib/coala_main.py [hunk 3]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
...
         args.apply_patch = False
-
```
Comment (ishanSrt) [inline coalib/coala_main.py]: same goes for this

**Comment by ishanSrt** [comment]
@Naveenaidu can you add a brief comment explaining your changes? will make the review faster. Also seems strange, since you are doing changes to `coala_main.py`, which means coala should not be working with any tag, but it is.

File: coalib/coala_main.py [hunk 4]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
     """
     all_actions_possible = provide_all_actions()
     apply_single = None
+
     if getattr(args, 'single_action', None) is not None:
         while True:
             for i, action in enumerate(all_actions_possible, 1):
                 console_printer.print(format_lines('{}'.format(
-                    action), symbol='['))
+                    action), symbol='[', line_nr=' ' + str(i) + '. '))
```
Comment (ishanSrt) [inline coalib/coala_main.py]: this doesn't seem right. the closing square bracket will never be added this way

File: coalib/coala_main.py [hunk 5]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
...
 
             line = format_lines(STR_ENTER_NUMBER, symbol='[')
-
             choice = input(line)
+            choice = str(choice)
```
Comment (ishanSrt) [inline coalib/coala_main.py]: you need to take an input over here. This is definitely wrong

File: coalib/coala_main.py [hunk 6]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
...
 
-            if choice.isalpha():
+            if choice.isalnum():
                 choice = choice.upper()
                 choice = '(' + choice + ')'
-                if choice == '(N)':
+                if choice == '(N)' or choice == '1':
```
Comment (ishanSrt) [inline coalib/coala_main.py]: Do Nothing is for 0 not 1

File: coalib/coala_main.py (hunk 4 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: The line_nr was for printing the option number on the screen.

Comment (ishanSrt) [reply coalib/coala_main.py]: >Also seems strange, since you are doing changes to coala_main.py, which means coala should not be working with any tag, but it is.

also I think the `action` handles that perfectly fine.

File: coalib/coala_main.py (hunk 6 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: Sorry! I'll change that.

File: coalib/coala_main.py (hunk 4 above)
Comment (ishanSrt) [reply coalib/coala_main.py]: ok that's correct

File: coalib/coala_main.py [hunk 7]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
...
-                    action), symbol='['))
+                    action), symbol='[', line_nr=' ' + str(i) + '. '))
 
             line = format_lines(STR_ENTER_NUMBER, symbol='[')
-
             choice = input(line)
+            choice = str(choice)
 
-            if choice.isalpha():
+            if choice.isalnum():
```
Comment (ishanSrt) [inline coalib/coala_main.py]: you don't need to change this

File: coalib/coala_main.py (hunk 5 above)
Comment (ishanSrt) [inline coalib/coala_main.py]: you need to take input over here.

File: coalib/coala_main.py (hunk 7 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: But then the user can select the options only when he enters letter like N for Do (0)Nothing,O etc.

File: coalib/coala_main.py [hunk 8]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
...
                 choice = choice.upper()
                 choice = '(' + choice + ')'
-                if choice == '(N)':
+                if choice == '(N)' or choice == '1':
                     apply_single = 'Do (N)othing'
                     break
                 for i, action in enumerate(all_actions_possible, 1):
-                    if choice in action:
+                    index = all_actions_possible.index(action)
```
Comment (ishanSrt) [inline coalib/coala_main.py]: isn't this equal to `i` now?

File: coalib/coala_main.py (hunk 5 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: My intention here was only to convert the  integer input we get into a string.

File: coalib/coala_main.py (hunk 8 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: Yeah sorry! I completly forgot about the enumerate functionality. I'll change that.

File: coalib/coala_main.py (hunk 7 above)
Comment (ishanSrt) [inline coalib/coala_main.py]: LGTM change (sorry I think I knocked my skull today 😛)

Comment (Naveenaidu) [reply coalib/coala_main.py]: No it's completely fine. :smiley: .

Comment (Naveenaidu) [reply coalib/coala_main.py]: So is the code fine?

Comment (ishanSrt) [reply coalib/coala_main.py]: yes

File: coalib/coala_main.py (hunk 6 above)
Comment (ishanSrt) [reply coalib/coala_main.py]: if you enter two numbers/options while selecting the option, coala currently performs the same action in that order. can you change your code so that the nature is preserved for now?

File: coalib/coala_main.py [hunk 9]
```diff
@@ -100,34 +100,35 @@ def run_coala(console_printer=None,
...
 
             line = format_lines(STR_ENTER_NUMBER, symbol='[')
-
             choice = input(line)
+            choice = str(choice)
 
-            if choice.isalpha():
+            if choice.isalnum():
                 choice = choice.upper()
                 choice = '(' + choice + ')'
```
Comment (ishanSrt) [inline coalib/coala_main.py]: better to change this to `choice = '(' + choice + ')' if choice.isalpha()` will save you the trouble for comparing of numeric characters later

**Comment by ishanSrt** [comment]
@Naveenaidu I looked at the `all_actions_possible`, the digits seem to be missing there too. So this won't work if the user enters a digit as an option. You need to perform some additional changes.

File: coalib/coala_main.py (hunk 6 above)
Comment (ishanSrt) [reply coalib/coala_main.py]: but since you need to apply single action for now, this seems to be correct for now. Wait, will verify and tell you.

**Comment by ishanSrt** [comment]
Also write some tests

File: coalib/coala_main.py (hunk 6 above)
Comment (ishanSrt) [reply coalib/coala_main.py]: umm.. just perform a check for the options user provides, for eg. `A1` should be claimed as an invalid option

File: coalib/coala_main.py (hunk 4 above)
Comment (ishanSrt) [reply coalib/coala_main.py]: https://gitter.im/coala/coala?at=5a8ed24ba2194eb80d9b7aa6 the output with `-A` tag seems to be showing 1 as Do Nothing, can you decrement the indices by 1. Also a * is printed before a 0 and you have added just a space before str(i) so make sure its well indented 👍

**Comment by Naveenaidu** [comment]
@ishanSrt  regarding the all_actions_possible . I tested it with an integer and it sure selects an action assosciated with the number in the action list.

**Comment by ishanSrt** [comment]
@Naveenaidu seems strange, `all_actions_possible` doesn't seem to be having numbers. What is it matching against?

File: coalib/coala_main.py (hunk 6 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: So i need a print a statement saying that it's an invalid option right?

Comment (ishanSrt) [reply coalib/coala_main.py]: something like line 127-129 of original file

**Comment by Naveenaidu** [comment]
@ishanSrt  All the actions are provied to all_actions_possible from provide_all_actions() as a list. And i guess the order of the list doen't change in python untill explicitly defined. Hence the actions can actually be referenced by the number. like all_actions_possible[0] gives us the Do (N)othing action , all_actions_possible[1] gives us the (S)how Patch Action.

**Comment by ishanSrt** [comment]
>I looked at the all_actions_possible, the digits seem to be missing there too

ok so compare with i in case of numerical input

**Comment by Naveenaidu** [comment]
@ishanSrt Should i move the if condition inside the try block. Because if its outside it's exiting with a error message when ctlr+D is used. But inside the try block there is a base exception case which handles the interrupt gracefully.

**Comment by ishanSrt** [comment]
@Naveenaidu yes even I saw that. But don't do it here. Instead file a new issue.

**Comment by Naveenaidu** [comment]
Okay! :+1:

**Comment by Naveenaidu** [comment]
@ishanSrt Can you please have a look at the PR.

**Comment by ishanSrt** [comment]
Sorry tomorrow but squash your commits into 1 and remove unnecessary additions and deletions of blank lines

**Comment by Naveenaidu** [comment]
@ishanSrt Okay :smile:

**Comment by aashraybhandar1** [comment]
@Naveenaidu  The build is failing coz there is no test written for the code you have added in the commit

**Comment by Naveenaidu** [comment]
@aashraybhandar1 Am i supposed to write a test for it?

**Comment by aashraybhandar1** [comment]
Also I don't think a colon is required after closes in your commit body

**Comment by Naveenaidu** [comment]
Yeah! I'll remove the colon. :+1:

**Comment by aashraybhandar1** [comment]
Your changes are in the `run_coala` method.See how tests for that method is written .You will find tests in the tests folder . The tests should be written in such a way that every line in your code should be run after the execution of the tests.If that doesn't happen the 'code coverage' is not 100% and the build fails.That's what's happening in your case. There is no test right now which tests line 104-112 in the coala_main file.

**Comment by aashraybhandar1** [comment]
But you didn't add any new lines just made changes to the existing code. So those tests were missing since beforehand. @ishanSrt  do you think the missing tests should be written?

**Comment by ishanSrt** [comment]
You should definitely add tests if you find them missing. Writing tests doesn't hurt anyone and prevents scenarios like these from happening 👍. You  see that coala doesn't have 100% code coverage from pytest.

**Comment by aashraybhandar1** [comment]
👍  Wasn't sure if they should be part of this PR or another should have been created. @Naveenaidu  try writing tests. Comment if you still have any doubts regarding them 😄

**Comment by Naveenaidu** [comment]
@aashraybhandar1 @ishanSrt  Am sorry to trouble you, but i have gone through the docs of coala about writing test and also through the tests folder of coala. It's going over my head :cry: . Can you please give me some pointers on which i can work through so that i can pass the tests. And also Travis and circle ci are failing due to an indentation fault. But when i run pytest locally all it shows is the failure in difftest. Can you please have a look at it. The coverage report is coming as 99% when i run pytest locally.

**Comment by ishanSrt** [comment]
@Naveenaidu maybe you need a unittest tutorial

File: coalib/coala_main.py [hunk 10]
```diff
@@ -104,27 +103,28 @@ def run_coala(console_printer=None,
...
                 for i, action in enumerate(all_actions_possible, 1):
-                    if choice in action:
+                    if choice in action or choice in str(i-1):
                         apply_single = action
                         break
+
                 if apply_single:
                     break
                 console_printer.print(format_lines(
-                                    'Please enter a valid letter.',
```
Comment (ishanSrt) [inline coalib/coala_main.py]: I don't think you need this much indentation over here. That's why the error

File: coalib/coala_main.py [hunk 11]
```diff
@@ -64,7 +64,6 @@ def run_coala(console_printer=None,
     """
     This is a main method that should be usable for almost all purposes and
     reduces executing coala to one function call.
-
```
Comment (ishanSrt) [inline coalib/coala_main.py]: please remove this unnecessary change

File: coalib/coala_main.py [hunk 12]
```diff
@@ -104,27 +103,28 @@ def run_coala(console_printer=None,
...
+                choice = '(' + choice + ')' if choice.isalpha() else choice
+                if choice == '(N)' or choice == '0':
                     apply_single = 'Do (N)othing'
                     break
                 for i, action in enumerate(all_actions_possible, 1):
-                    if choice in action:
+                    if choice in action or choice in str(i-1):
                         apply_single = action
                         break
+
```
Comment (ishanSrt) [inline coalib/coala_main.py]: this is also not required

**Comment by Naveenaidu** [comment]
And also when i ran pytest on the master, it still gives me the same error :- 
FAIL tests/results/DiffTest.py::DiffTest::test_from_clang_fixit

**Comment by Naveenaidu** [comment]
The checks for the user input are done.

**Comment by ishanSrt** [comment]
>The checks for the user input are done.

I don't see them 😲

**Comment by Naveenaidu** [comment]
Sorry! One second i forgot to push them.

File: coalib/coala_main.py [hunk 13]
```diff
@@ -104,27 +103,28 @@ def run_coala(console_printer=None,
...
                 if apply_single:
                     break
                 console_printer.print(format_lines(
-                                    'Please enter a valid letter.',
-                                    symbol='['))
+                    'Please enter a valid letter or number.',
```
Comment (ishanSrt) [inline coalib/coala_main.py]: Lets not change this error message for now. Maybe in a different issue.

**Comment by Naveenaidu** [comment]
@ishanSrt  Can you please have look at the code now!

File: coalib/coala_main.py [hunk 14]
```diff
@@ -104,27 +103,27 @@ def run_coala(console_printer=None,
...
+                    if choice in action or choice in str(i-1):
                         apply_single = action
                         break
+
                 if apply_single:
                     break
-                console_printer.print(format_lines(
-                                    'Please enter a valid letter.',
-                                    symbol='['))
+            console_printer.print(format_lines(
```
Comment (ishanSrt) [inline coalib/coala_main.py]: this doesn't seem correct

**Comment by ishanSrt** [comment]
still no check for user input

File: coalib/coala_main.py [hunk 15]
```diff
@@ -104,27 +103,27 @@ def run_coala(console_printer=None,
...
+                'Please enter a valid letter or number.', symbol='['))
```
Comment (ishanSrt) [inline coalib/coala_main.py]: don't change this message for now. Open a different issue, if you feel its not descriptive enough. More newcomer issues that way 😉

File: coalib/coala_main.py (hunk 14 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: May i know why? When the user doesn't enter a number or a alphabet it comes out of the if choice.isaplha() and prints the error message.

File: coalib/coala_main.py (hunk 15 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: Okay! I'll keep that in mind and open a issue as soon as this PR gets merged.

**Comment by Naveenaidu** [comment]
There is no need to check for user input as it automatically comes out of for  the loop when the user enter more than one letter or number or a combination of both. I checked all of thoose conditons locally.

File: coalib/coala_main.py (hunk 14 above)
Comment (ishanSrt) [reply coalib/coala_main.py]: won't this be printed either way whether the the condition is true or false?

**Comment by ishanSrt** [comment]
>There is no need to check for user input as it automatically comes out of for the loop when the user enter more than one letter or number or a combination of both. I checked all of thoose conditons locally.

don't you need to print error message in such case?

**Comment by Naveenaidu** [comment]
It prints it. Console.printer is outside the if choice.isaplha(), Suppose a user enter a combination of alphabets or numbers, it goes inside the if condition but since it fails all the comparison inside the 
for i, action in enumerate(all_actions_possible, 1): it breaks from the for loop and comes out of the if condition. And as soon as it exits the if condition it encounters with the console.print statement which will print the error message,

**Comment by Naveenaidu** [comment]
And for the cases where there are combinations of both alphabets and a number it fails the 
if choice.isaplha() and encounter the console.print statement which prints the error message.

File: coalib/coala_main.py (hunk 14 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: It seems strange, even i thought that. But it doen't print the error message when a correct option is given.

Comment (Naveenaidu) [reply coalib/coala_main.py]: The output looks like this :-
coala --bears PEP8Bear,SpaceConsistencyBear,PyCommentedBear,PycodestyleBear --files coala_main.py -A
[    ] 0. Do (N)othing
[    ] 1. (S)how patch
[    ] 2. (A)pply patch
[    ] 3. Add (I)gnore comment
[    ] 4. (O)pen file
[    ] 5. Print Aspec(T) Information
[    ] 6. Print (D)ebug message
[    ] 7. Print (M)ore info
[    ] Enter number (Ctrl-D to exit): 34
[    ] Please enter a valid letter or number.
[    ] 0. Do (N)othing
[    ] 1. (S)how patch
[    ] 2. (A)pply patch
[    ] 3. Add (I)gnore comment
[    ] 4. (O)pen file
[    ] 5. Print Aspec(T) Information
[    ] 6. Print (D)ebug message
[    ] 7. Print (M)ore info
[    ] Enter number (Ctrl-D to exit): 1
[WARNING][13:21:01] Default coafile '.coafile' not found!
Here's what you can do:
* add `--save` to generate a config file with your current options
* add `-I` to suppress any use of config files

[WARNING][13:21:01] No bears matching 'PyCommentedBear' were found. Make sure you have coala-bears installed or you have typed the name correctly.
[WARNING][13:21:01] acquire_settings: section parameter is deprecated.
Please enter a value for the setting "use_spaces" (True if spaces are to be used instead of tabs.) needed by SpaceConsistencyBear for section "cli": 
True
Executing section cli...

coala_main.py
[   9] from·coalib.processes.Processing·import·execute_section,·simplify_section_result
**** PycodestyleBear (E501) [Section: cli | Severity: NORMAL] ****
!    ! E501 line too long (80 > 79 characters)

Comment (Naveenaidu) [reply coalib/coala_main.py]: @ishanSrt  It won't print the console.print statement every time because, we are checking for
" if apply_single: break" before printing the error message. And in case the user enters the correct option apply_single becomes true and it breaks from the while loop. If the user doen't then that if conditon fails and it comes to the error message part.

**Comment by ishanSrt** [comment]
>won't this be printed either way whether the the condition is true or false?

what about this?

**Comment by ishanSrt** [comment]
Ok then, it seems good.

**Comment by ishanSrt** [comment]
Take care of this: https://github.com/coala/coala/pull/5198#discussion_r169995771

File: coalib/coala_main.py (hunk 4 above)
Comment (Naveenaidu) [reply coalib/coala_main.py]: Yeah! I'll keep that in mind.

Comment (Naveenaidu) [reply coalib/coala_main.py]: @ishanSrt  Is the * necessary to be added to 0 for Do (N)othing. What if we open a new issue for it? :smile:

**Comment by Naveenaidu** [comment]
@ishanSrt Should a new issue be opened for writing tests for the PR. And also when this PR gets merged there will be a lot of issues such as:-
1)Change the error message when the user inputs a incorrect option.
2)To remove the error message when ctlr+D is pressed.
3)Adding colors to the output of the options.
4)The print_beginning() section appears only after an option is selected but it should appear as soon as a coala command is executed with -A tag.

I think it would be better we also add a writing a test case to the issue list. :smile:

**Comment by ishanSrt** [comment]
No you should write tests for whatever you are doing in the same PR.

**Comment by Naveenaidu** [comment]
Okay! I'll get to it then :smile: . I just had one more doubt, should i add a test case to any existing file or should a new test file be created to handle it. If it should be added to an existing file, can you please refer me to that file. I searched the test folder but couldn't find anything. :sweat_smile: .

**Comment by aashraybhandar1** [comment]
Existing test file. Check the test folder there is a file where the run_coala method is being tested. You have edited that method so you have to write it in the same file. Check test folder once more 😉

**Comment by monsij** [comment]
This issue is a bug.Shouldn't it be Fixes instead of Closes?

**Comment by Naveenaidu** [comment]
@ishanSrt Can you have look at the PR now. I made the changes as you suggested. Moved the printing of options inside ConsoleInteraction.py . I'll squash the commits  if you think the code is fine :wink: . I did think about it earlier but decided i'll open up an issue when this PR get's merged :laughing: .

**Comment by nvzard** [comment]
@Naveenaidu you need 100% test coverage to pass the builds, write tests for `\output\ConsoleInteraction.py  677-678, 683, 673->683, 676->677, 679->683`, also please make the changes suggested by Gitmate ;)

**Comment by nvzard** [comment]
And you should not push changes from your master-branch :o

**Comment by Naveenaidu** [comment]
@nvzard I am in the process of writing test, it just taking a little longer to understand how to write test :sweat_smile:  .  And yes I'll keep in mind not to push changes from the master branch. Thanks for the advice. :smile:

File: coalib/output/ConsoleInteraction.py [hunk 16]
```diff
@@ -642,11 +654,40 @@ def choose_action(console_printer, actions, apply_single=False):
...
+                choice = '(' + choice + ')' if choice.isalpha() else choice
+                if choice == '(N)' or choice == '0':
+                    APPLY_SINGLE_ACTION = 'Do (N)othing'
+                    return ('Do (N)othing', 'Do (N)othing')
+                for i, action in enumerate(all_actions_possible, 0):
+                    if choice in action or choice in str(i):
+                        APPLY_SINGLE_ACTION = action
+                        return (action,action)
+            console_printer.print(format_lines(
+                'Please enter a valid letter or number.', symbol='['))
```
[the remark this answers is not in the document]
Comment (ishanSrt) [reply coalib/output/ConsoleInteraction.py]: I think you can reuse the code in the present lines 694-728 by subtle changes

File: coalib/output/ConsoleInteraction.py [hunk 17]
```diff
@@ -93,6 +94,8 @@ def highlight_text(no_color, text, style, lexer=TextLexer()):
                ShowAppliedPatchesAction(),
                GeneratePatchesAction())
 DIFF_EXCERPT_MAX_SIZE = 4
+APPLY_SINGLE = 1
+APPLY_SINGLE_ACTION = 'NONE'
```
Comment (ishanSrt) [inline coalib/output/ConsoleInteraction.py]: hmmmm... or you can modify `Processing.py` so that the responsible queue of methods which actually call `choose_action` gets called only once if `apply_single` is True ov🤔er there

File: coalib/output/ConsoleInteraction.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: Okay! I'll work on it.  :smile:

File: coalib/output/ConsoleInteraction.py (hunk 17 above)
Comment (ishanSrt) [reply coalib/output/ConsoleInteraction.py]: seems better than maintaining these global variables

Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: Well, choose_action is actually called the number of times the patches needed to be applied, if i make changes in the processing.py to call choose_action only once then only patch will get applied.

File: coalib/output/ConsoleInteraction.py [hunk 18]
```diff
@@ -625,6 +628,15 @@ def get_action_info(section, action, failed_actions):
 
     return action.name, section
 
+def provide_all_actions():
```
Comment (ishanSrt) [inline coalib/output/ConsoleInteraction.py]: Do you need this method inspite of having access to the param `actions`? 🧐

**Comment by ishanSrt** [comment]
As per your making your PRs from the master branch, you may eventually need to close this and re-open it from a different branch as that's one of the very first things mentioned in the newcomer's guide. Just make sure you reference this over there.

File: coalib/output/ConsoleInteraction.py (hunk 18 above)
Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: I needed advice on that. I was unsure if i need to keep that method or not. Should i let the apply_single method display what the original author intended it to or should i change it.

Comment (ishanSrt) [reply coalib/output/ConsoleInteraction.py]: You mean modifying the doctoring?

**Comment by Naveenaidu** [comment]
Sorry! I'll keep that in mind.

File: coalib/output/ConsoleInteraction.py (hunk 17 above)
Comment (ishanSrt) [reply coalib/output/ConsoleInteraction.py]: but you have access to `apply_single` in `Processing.py` right?

Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: Yes we do. But it doesn't control how the patches are applied i think. I don't know i might be wrong. But as much as i experimented, irrespective of whether it is a apply_single action or not, choose_action() method is being called everytime a patch has to be applied. choose_action method is also responsible for passing the selected action back to processing.py.

Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: When apply_single is selected the action is stored in a variable and is checked with the available actions everytime in the choose_Action and passes it back to process. If we want to change that then i think we need to write a whole new function for it and make lots of changes :sweat_smile:

File: coalib/output/ConsoleInteraction.py (hunk 18 above)
Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: Do you remember the output you posted while filing for the issue. The ouput was actually coded into the print statments . But if you see choose_action() method, it does some  magic to get only the required actions and print them.

Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: I think the reason thoose function were added was to maybe add an extra functionality later on. I think it's better to keep it as it is. What's your opinion?

File: coalib/output/ConsoleInteraction.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (aashraybhandar1) [reply coalib/output/ConsoleInteraction.py]: I think global variables should be declared first

[the remark this answers is not in the document]
Comment (Naveenaidu) [reply coalib/output/ConsoleInteraction.py]: @aashraybhandar1  The global variables have been declared at the beginning of the code. Please check the complete version of code.

[the remark this answers is not in the document]
Comment (aashraybhandar1) [reply coalib/output/ConsoleInteraction.py]: `action_desc` and `actions_name` are declared earlier

**Comment by adtac** [comment]
@ishanSrt I don't think it was necessary to close this PR. Just finish the PR from the master branch and remember to not use it next time.

**Comment by Naveenaidu** [comment]
@adtac it was me who closed the PR not @ishanSrt . Sorry! I'll keep that in mind from next time.
